saliency/xrai.py

```diff
@@ -0,0 +1,555 @@
+        segs: Input segments as a list of boolean masks. XRAI uses these to
+              compute attribution sums.
+        gain_fun: The function that computes XRAI area attribution from source
+                  attributions. Defaults to _gain_density, which calculates the
+                  density of attributions in a mask.
+        area_perc_th: This parameter is ignored. Fast version always computes
+                      to 1.0. It is left here for API compatibility.
+        verbose: 1 prints information about every added mask. >2 prints
+                 information about skipped masks due to min_pixel_diff.
+                 Defaults to 0, which is silent.
+        min_pixel_diff: Do not consider masks that have difference less than
+                        this number compared to the current mask. Set it to 1
+                        to remove masks that completely overlap with the
+                        current mask.
+        integer_segments: See XRAIParameters. Defaults to True.
+
+    Returns:
+        tuple: saliency heatmap and list of masks or an integer image with
+               area ranks depending on the parameter integer_segments.
+    """
+    output_attr = -np.inf * np.ones(shape=attr.shape, dtype=np.float)
+
+    n_masks = len(segs)
+    current_mask = np.zeros(attr.shape, dtype=bool)
+
+    masks_trace = []
+
+    # Sort all masks based on gain, ignore overlaps
+    seg_attrs = [gain_fun(seg_mask, attr) for seg_mask in segs]
+    segs, seg_attrs = list(zip(
+        *sorted(zip(segs, seg_attrs), key=lambda x: -x[1])))
+
+    for i, added_mask in enumerate(segs):
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+       # If mask does not add more than min_pixel_diff to current mask, skip
+      mask_pixel_diff = _get_diff_cnt(added_mask, current_mask)
+      if mask_pixel_diff < min_pixel_diff:
+        if verbose > 2:
+          logging.info("Skipping mask with pixel difference: {:.3g},".format(
+              mask_pixel_diff))
+        continue
+      mask_gain = gain_fun(mask_diff, attr)
+      masks_trace.append((mask_diff, mask_gain))
+      output_attr[mask_diff] = mask_gain
+      current_mask = np.logical_or(current_mask, added_mask)
+      if verbose:
+        current_attr_sum = np.sum(attr[current_mask])
+        current_area_perc = np.mean(current_mask)
+        logging.info("{} of {} masks processed,"
+                     "attr_sum: {}, area: {:.3g}/{:.3g}".format(
+                         i + 1, n_masks, current_attr_sum, current_area_perc,
+                         area_perc_th))
+    masks_trace = list(zip(*sorted(masks_trace, key=lambda x: -x[1])))[0]
+    if integer_segments:
+      attr_ranks = np.zeros(shape=attr.shape, dtype=np.int)
+      for i, mask in enumerate(masks_trace):
+        attr_ranks[mask] = i + 1
+      attr_ranks[attr_ranks==0] = np.max(attr_ranks) + 1
+      return output_attr, attr_ranks
+    else:
+      return output_attr, masks_trace
```

Comment:
Same here. Should the the union of the excluded masks be included?

Reply:
done.